platform/features/timeline/src/actions/TimelineCSVExporter.js

```diff
@@ -0,0 +1,156 @@
+ * http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * Open MCT Web includes source code licensed under additional open source
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+/*global define,Promise*/
+
+define([
+    "./IdColumn",
+    "./ModeColumn",
+    "./CompositionColumn",
+    "./MetadataColumn",
+    "./TimespanColumn"
+], function (
+    IdColumn,
+    ModeColumn,
+    CompositionColumn,
+    MetadataColumn,
+    TimespanColumn
+) {
+    'use strict';
+
+    /**
+     * A description of how to populate a given column within a
+     * prepared table of domain object data, for CSV export.
+     * @interface platform/features/timeline.TimelineCSVColumn
+     */
+
+    /**
+     * Get the value that belongs in this column for a given
+     * domain object.
+     * @memberof {platform/features/timeline.TimelineCSVColumn#}
+     * @method value
+     * @param {DomainObject} domainObject the domain object
+     *        represented by this row
+     * @returns {string|Promise<string>} the value for this cell
+     */
+
+    /**
+     * Get the name of this column, as belongs in a header.
+     * @memberof {platform/features/timeline.TimelineCSVColumn#}
+     * @method name
+     * @returns {string} the name of this column
+     */
+
+    /**
+     * Handles conversion of a list of domain objects to a table
+     * representation appropriate for CSV export.
+     *
+     * @param {DomainObject[]} domainObjects the objects to include
+     *        in the exported data
+     * @constructor
+     * @memberof {platform/features/timeline}
+     */
+    function TimelineCSVExporter(domainObjects) {
+        var maxComposition = 0,
+            maxRelationships = 0,
+            columnNames = {},
+            columns = [],
+            foundTimespan = false,
+            i;
+
+        function addMetadataProperty(property) {
+            var name = property.name;
+            if (!columnNames[name]) {
+                columnNames[name] = true;
+                columns.push(new MetadataColumn(name));
+            }
+        }
+
+        columns.push(new IdColumn());
+
+        domainObjects.forEach(function (domainObject) {
+            var model = domainObject.getModel(),
+                compositionLength = model.composition ?
+                    model.composition.length : 0,
+                relationshipLength = (model.relationships || {}).modes ?
```

Comment:
Mixing binary and ternary operators here makes it a little difficult to follow. Could either do it in two steps (define relationships, then relationshipLength, or alternatively do something like -
``` javascript
var model = domainObject.getModel(),
    composition = model.composition || [],
    modes = (model.relationships || {}).modes || [];
    ...
    maxComposition = Math.max(maxComposition, composition.length);
    maxRelationships = Math.max(maxRelationships, relationships.length);
```

I'm nitpicking though.